tests/image/dataframe_iterator_test.py

```diff
@@ -0,0 +1,653 @@
+    # create iterator
+    generator = image_data_generator.ImageDataGenerator(
+        validation_split=validation_split
+    )
+    df_sparse_iterator = generator.flow_from_dataframe(df,
+                                                       str(tmpdir),
+                                                       class_mode="sparse")
+    if np.isnan(next(df_sparse_iterator)[:][1]).any():
+        raise ValueError('Invalid values.')
+
+    with pytest.raises(ValueError):
+        generator.flow_from_dataframe(
+            df, tmpdir, subset='foo')
+
+    train_iterator = generator.flow_from_dataframe(df, str(tmpdir),
+                                                   subset='training')
+    assert train_iterator.samples == num_training
+
+    valid_iterator = generator.flow_from_dataframe(df, str(tmpdir),
+                                                   subset='validation')
+    assert valid_iterator.samples == count - num_training
+
+    # check number of classes and images
+    assert len(train_iterator.class_indices) == num_classes
+    assert len(train_iterator.classes) == num_training
+    assert len(set(train_iterator.filenames) &
+               set(filenames)) == num_training
+
+
+def test_dataframe_iterator_with_custom_indexed_dataframe(all_test_images, tmpdir):
+    num_classes = 2
+
+    # save the images in the tmpdir
+    count = 0
+    filenames = []
+    for test_images in all_test_images:
+        for im in test_images:
+            filename = "image-{}.png".format(count)
+            filenames.append(filename)
+            im.save(str(tmpdir / filename))
+            count += 1
+
+    # create dataframes
+    classes = np.random.randint(num_classes, size=len(filenames))
+    classes = [str(c) for c in classes]
+    df = pd.DataFrame({"filename": filenames,
+                       "class": classes})
+    df2 = pd.DataFrame({"filename": filenames,
+                        "class": classes},
+                       index=np.arange(1, len(filenames) + 1))
+    df3 = pd.DataFrame({"filename": filenames,
+                        "class": classes},
+                       index=filenames)
+
+    # create iterators
+    seed = 1
+    generator = image_data_generator.ImageDataGenerator()
+    df_iterator = generator.flow_from_dataframe(
+        df, str(tmpdir), seed=seed)
+    df2_iterator = generator.flow_from_dataframe(
+        df2, str(tmpdir), seed=seed)
+    df3_iterator = generator.flow_from_dataframe(
+        df3, str(tmpdir), seed=seed)
+
+    # Test all iterators return same pairs of arrays
+    for _ in range(len(filenames)):
+        a1, c1 = next(df_iterator)
+        a2, c2 = next(df2_iterator)
+        a3, c3 = next(df3_iterator)
+        assert np.array_equal(a1, a2)
+        assert np.array_equal(a1, a3)
+        assert np.array_equal(c1, c2)
+        assert np.array_equal(c1, c3)
+
+
+def test_dataframe_iterator_n(all_test_images, tmpdir):
+
+    # save the images in the tmpdir
+    count = 0
```

Comment:
I'm not sure of the value of this test?

Reply:
I'm also not sure, it was already there. @Vijayabhaskar96  do you know? I think you originally made  this one.

Reply:
Looks like It matches the number of filenames found and the number of files in the dataframe is the same. @rragundez I think I didn't author this.

Reply:
thanks @Vijayabhaskar96